### PR TITLE
[SPARK-49115][DOCS] Add docs for state metadata source for operators using schema format v2 such as transformWithState

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Add docs for state metadata source for operators using schema format v2 such as transformWithState

### Why are the changes needed?
Adding documentation for state metadata source and its use with newer operators such as `transformWithState`

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
Existing tests

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by anishshri-db** [comment]
cc - @HeartSaVioR - PTAL, thx !

File: docs/structured-streaming-state-data-source.md [hunk 1]
```diff
@@ -264,13 +264,14 @@ The output schema will also be different from the normal output.
...
 
-## State metadata source
+## State Metadata Source
 
 Before querying the state from existing checkpoint via state data source, users would like to understand the information for the checkpoint, especially about state operator. This includes which operators and state store instances are available in the checkpoint, available range of batch IDs, etc.
 
 Structured Streaming provides a data source named "State metadata source" to provide the state-related metadata information from the checkpoint.
 
 Note: The metadata is constructed when the streaming query is running with Spark 4.0+. The existing checkpoint which has been running with lower Spark version does not have the metadata and will be unable to query/use with this metadata source. It is required to run the streaming query pointing the existing checkpoint in Spark 4.0+ to construct the metadata before querying.
+For operators using the operator metadata format version 1, the metadata is written once and does not change. For operators using metadata format version 2, this metadata can change and can be queried by providing the relevant batchId.
```
Comment (HeartSaVioR) [inline docs/structured-streaming-state-data-source.md:274]: We don't expect users to know about the operator metadata format, which operator uses the version 1 and which operator doesn't.

Flipping the coin, would there be a practice which users can simply follow without knowing the details? Shall we just guide users to always specify the batch ID if they want to know the metadata in "point of time", regardless of format version? My understanding is that the outcome should be the same.

File: docs/structured-streaming-state-data-source.md [hunk 2]
```diff
@@ -344,6 +368,11 @@ Each row in the source has the following schema:
   <td>int</td>
   <td>The maximum batch ID available for querying state. The value could be invalid if the streaming query taking the checkpoint is running, as the query will commit further batches.</td>
 </tr>
+<tr>
+  <td>operatorProperties</td>
+  <td>string</td>
+  <td>List of properties used by the operator encoded as JSON. Available only for operators using schema format version 2.</td>
```
Comment (HeartSaVioR) [inline docs/structured-streaming-state-data-source.md:374]: Let's not expose the format version - we can just say "operator dependent".

File: docs/structured-streaming-state-data-source.md [hunk 3]
```diff
@@ -310,6 +311,29 @@ Dataset<Row> df = spark
...
+
+The following configurations are optional:
+
+<table>
+<thead><tr><th>Option</th><th>Value</th><th>Default</th><th>Meaning</th></tr></thead>
+<tr>
+  <td>batchId</td>
+  <td>numeric value</td>
+  <td>Last committed batch if available, else 0</td>
+  <td>Optional batchId used to retrieve operator metadata at that batch. Only applicable for operators using schema format version 2.</td>
```
Comment (HeartSaVioR) [inline docs/structured-streaming-state-data-source.md:333]: Let's just hide the fact it's effective only for format version 2. Users wouldn't need to know about it and the expected behavior would be the same.

**Review by HeartSaVioR — COMMENTED** [review]
Thanks for the work! My general suggestion is that we need to document in a way users would understand. They aren't expected to know about internals.

File: docs/structured-streaming-state-data-source.md (hunk 1 above)
Comment (anishshri-db) [reply docs/structured-streaming-state-data-source.md]: Done

File: docs/structured-streaming-state-data-source.md (hunk 2 above)
Comment (anishshri-db) [reply docs/structured-streaming-state-data-source.md]: Done

File: docs/structured-streaming-state-data-source.md (hunk 3 above)
Comment (anishshri-db) [reply docs/structured-streaming-state-data-source.md]: Done

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**